server/src/main/java/io/crate/expression/operator/all/AllEqOperator.java

```diff
@@ -0,0 +1,129 @@
+import io.crate.expression.scalar.ArrayUnnestFunction;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
+import io.crate.lucene.LuceneQueryBuilder;
+import io.crate.metadata.FunctionType;
+import io.crate.metadata.Reference;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.sql.tree.ComparisonExpression;
+import io.crate.types.TypeSignature;
+
+public class AllEqOperator extends AllOperator {
+
+    public static final String NAME = OPERATOR_PREFIX + ComparisonExpression.Type.EQUAL.getValue();
+    public static final Signature SIGNATURE = Signature.builder(AllEqOperator.NAME, FunctionType.SCALAR)
+            .argumentTypes(TypeSignature.parse("E"), TypeSignature.parse("array(E)"))
+            .returnType(Operator.RETURN_TYPE.getTypeSignature())
+            .typeVariableConstraints(typeVariable("E"))
+            .features(Feature.DETERMINISTIC, Feature.STRICTNULL)
+            .build();
+
+    public AllEqOperator(Signature signature, BoundSignature boundSignature) {
+        super(signature, boundSignature, result -> result == 0);
+    }
+
+    public static Query refMatchesAllArrayLiteral(Reference probe, List<?> values, LuceneQueryBuilder.Context context) {
+        // col = ALL ([1,2,3]) --> col=1 and col=2 and col=3
+        LinkedHashSet<?> uniqueValues = new LinkedHashSet<>(values);
+        if (uniqueValues.isEmpty()) {
+            return new MatchAllDocsQuery();
+        }
+        if (uniqueValues.contains(null)) {
+            return new MatchNoDocsQuery("Cannot match nulls");
+        }
+        if (uniqueValues.size() > 1) {
+            // `col=x and col=y` evaluates to `false` unless `x == y`
+            return new MatchNoDocsQuery("A single value cannot match more than one unique values");
+        }
+        var value = uniqueValues.getFirst();
+        // col = all([1]) --> col = 1
+        Function eq = new Function(
+            EqOperator.SIGNATURE,
+            List.of(probe, Literal.ofUnchecked(probe.valueType(), value)),
+            EqOperator.RETURN_TYPE
+        );
+        return context.visitor().visitFunction(eq, context);
+    }
+
+    public static Query literalMatchesAllArrayRef(Function allEq, LuceneQueryBuilder.Context context) {
+        // 1 = ALL(array_col)
+        // --> 1 = array_col[1] and 1 = array_col[2] and 1 = array_col[3]
+        // --> not(1 != array_col[1] or 1 != array_col[2] or 1 != array_col[3])
+        // --> not(1 != ANY(array_col))
+        Function notAnyNeq = new Function(
+            NotPredicate.SIGNATURE,
+            List.of(
+                new Function(
+                    AnyNeqOperator.SIGNATURE,
+                    allEq.arguments(),
+                    Operator.RETURN_TYPE
+                )
+            ),
+            Operator.RETURN_TYPE
+        );
+        return context.visitor().visitFunction(notAnyNeq, context);
```

Comment:
The previous implementation was flawed, this should be correct implemntation but unfortunately the performance improvement is not as much:
```
# Results (server side duration in ms)
V1: 5.9.0-8c5d5fa9168e253a340abd6fba604c27603e9275
V2: 5.9.0-ecc9d9e37e1ec50897c8384f907521e821874c69

Q: select count(*) from t where 1 = all(xs) -- xs hold 1M arrays(fixed size: 8) with ints between 1-10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      187.027 ±  342.611 |    124.763 |    127.714 |    131.938 |   3047.745 |
|   V2    |      160.090 ±  196.127 |    134.480 |    136.223 |    137.687 |   2168.458 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  15.52%                           +   6.45%   
There is a 66.49% probability that the observed difference is not random, and the best estimate of that difference is 15.52%
The test has no statistical significance
```

Reply:
This is because in the end we need a generic function filter to identify null containing arrays like `[1, null]`.

Reply:
Do we still need a release note entry?

Reply:
I think 15% is still noteworthy.

Reply:
Benchmark results of the newer approach [89e96e1](https://github.com/crate/crate/pull/16502/commits/89e96e129f3ea85d8e596a35f2d5a02b74709391):
```
# Results (server side duration in ms)
V1: 5.9.0-8f368832a8a35d590d50e140354f06845588c262
V2: 5.9.0-89e96e129f3ea85d8e596a35f2d5a02b74709391

Q: select count(*) from t where 1 = all(xs)
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      341.356 ±  324.351 |    278.163 |    283.603 |    290.526 |   2843.144 |
|   V2    |        5.838 ±   14.941 |      3.579 |      4.252 |      4.737 |    208.666 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 193.27%                           - 194.09%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 193.27%
The test has statistical significance
```